lib/hooks/userspace/hook-grpc.js

```diff
@@ -129,6 +128,299 @@ function makeClientConstructorWrap(makeClientConstructor) {
+      return func.apply(this, arguments);
+    }, context);
+  });
+}
+
+/**
+ * Wraps a server streaming function in order to record trace spans.
+ * @param {cls.Namespace} namespace The CLS namespace.
+ * @param {Object} handlerSet An object containing references to the function handle,
+ * as well as serialize and deserialize handles.
+ * @param {string} requestName The human-friendly name of the request.
+ */
+function wrapServerStream(namespace, handlerSet, requestName) {
+  var rootContext;
+  // CLS context object
+  var context = {};
+  // Start right before deserialization (which precedes call to function)
+  // End right after output stream 'finish' event
+  shimmer.wrap(handlerSet, 'deserialize', function (deserialize) {
+    return namespace.bind(function deserializeTrace() {
+      rootContext = startRootSpanForRequest(requestName);
+      return deserialize.apply(this, arguments);
+    }, context);
+  });
+  shimmer.wrap(handlerSet, 'func', function (func) {
+    return namespace.bind(function funcTrace(call) {
+      if (agent.config_.enhancedDatabaseReporting) {
+        shimmer.wrap(call, 'sendMetadata', sendMetadataWrapper(rootContext));
+      }
+      if (agent.config_.enhancedDatabaseReporting) {
+        rootContext.addLabel('argument', JSON.stringify(call.request));
+      }
+      var spanEnded = false;
+      var endSpan = function () {
+        if (!spanEnded) {
+          spanEnded = true;
+          endRootSpanForRequest(rootContext, requestName);
+        }
+      };
+      call.on('finish', function () {
+        // End the span unless there is an error.
+        if (call.status.code === 0) {
+          endSpan();
+        }
+      });
+      call.on('error', function (err) {
+        if (agent.config_.enhancedDatabaseReporting) {
+          rootContext.addLabel('error', err);
+        }
+        endSpan();
+      });
+      return func.apply(this, arguments);
+    }, context);
+  });
+}
+
+/**
+ * Wraps a client streaming function in order to record trace spans.
+ * @param {cls.Namespace} namespace The CLS namespace.
+ * @param {Object} handlerSet An object containing references to the function handle,
+ * as well as serialize and deserialize handles.
+ * @param {string} requestName The human-friendly name of the request.
+ */
+function wrapClientStream(namespace, handlerSet, requestName) {
+  var rootContext;
+  // CLS context object
+  var context = {};
+  // Start right before call to function
+  // End right after serialization (which is done in the callback)
+  shimmer.wrap(handlerSet, 'func', function (func) {
+    return namespace.bind(function funcTrace(stream, callback) {
+      rootContext = startRootSpanForRequest(requestName);
+      if (agent.config_.enhancedDatabaseReporting) {
+        shimmer.wrap(stream, 'sendMetadata', sendMetadataWrapper(rootContext));
+      }
+      // Preserve context in stream event handlers.
+      namespace.bindEmitter(stream);
+      // arguments[1] is the callback
```

Comment:
Are there any optional arguments that could cause the callback to be at a different position?